File: server/services/dataSourceClients/apolloClient.ts
import axios from 'axios';

export interface ApolloSearchParams {
  keywords?: string;
  industry?: string;
  location?: string;
  jobTitles?: string[];
  companySize?: string;
  technologies?: string[];
  page?: number;
  limit?: number;
}

export interface ApolloProspect {
  id: string;
  first_name: string;
  last_name: string;
  title: string;
  company: {
    name: string;
    industry: string;
    size: string;
    location: string;
    technologies?: string[];
  };
  email: string;
  phone?: string;
  linkedin_url?: string;
  verified_at?: string;
  confidence_score: number;
}

export class ApolloClient {
  private apiKey: string;
  private baseUrl = 'https://api.apollo.io/v1';

  constructor(apiKey: string) {
    this.apiKey = apiKey;
    console.log('🚀 Apollo Client initialized with API key:', apiKey ? 'Present' : 'Missing');
  }

  async search(params: ApolloSearchParams): Promise<{ results: ApolloProspect[], total: number }> {
    if (!this.apiKey || this.apiKey === 'mock') {
      // Return empty results if no valid API key
      console.log('❌ Apollo API key missing or set to mock');
      return { results: [], total: 0 };
    }

    try {
      // Build request payload with proper Apollo API structure
      const payload: any = {
        per_page: Math.min(params.limit || 50, 200), // Apollo max is 200 per page
        page: params.page || 1,
      };

      // Add search criteria - Apollo requires at least one search parameter
      if (params.keywords) {
        payload.q_keywords = params.keywords;
      }

      if (params.jobTitles && params.jobTitles.length > 0) {
        payload.person_titles = params.jobTitles;
      }

      if (params.industry) {
        payload.organization_industry_tag_names = [params.industry];
      }

      if (params.location) {
        payload.organization_locations = [params.location];
      }

      const companySizes = this.mapCompanySize(params.companySize);
      if (companySizes.length > 0) {
        payload.organization_num_employees_ranges = companySizes;
      }

      if (params.technologies && params.technologies.length > 0) {
        payload.technologies = params.technologies;
      }

      // Ensure we have at least one search parameter for Apollo API
      if (!payload.q_keywords && (!payload.person_titles || payload.person_titles.length === 0) && 
          (!payload.organization_industry_tag_names || payload.organization_industry_tag_names.length === 0)) {
        payload.q_keywords = "sales manager"; // Default fallback search
      }

      console.log('🚀 Apollo API request payload:', JSON.stringify(payload, null, 2));

      const response = await axios.post(
        `${this.baseUrl}/mixed_people/search`,
        payload,
        {
          headers: {
            'X-Api-Key': this.apiKey,
            'Content-Type': 'application/json',
          },
          timeout: 30000, // 30 second timeout
        }
      );

      console.log('✅ Apollo API response received:', {
        people_count: response.data.people?.length || 0,
        total_entries: response.data.pagination?.total_entries || 0,
        current_page: response.data.pagination?.page || 0,
        per_page: response.data.pagination?.per_page || 0
      });

      if (!response.data.people || !Array.isArray(response.data.people)) {
        console.error('❌ Apollo API returned invalid data structure:', response.data);
        return { results: [], total: 0 };
      }

      const prospects = response.data.people.map((person: any, index: number) => {
        try {
          return {
            id: person.id || `apollo_${Date.now()}_${index}`,
            first_name: person.first_name || '',
            last_name: person.last_name || '',
            title: person.title || '',
            company: {
              name: person.organization?.name || 'Unknown Company',
              industry: person.organization?.industry || '',
              size: this.formatCompanySize(person.organization?.estimated_num_employees),
              location: this.formatLocation(person.organization),
              technologies: person.organization?.technologies || [],
            },
            email: person.email || '',
            phone: person.phone_numbers?.[0]?.sanitized_number || person.phone_numbers?.[0]?.number,
            linkedin_url: person.linkedin_url || '',
            verified_at: person.email_verified_at,
            confidence_score: this.calculateConfidenceScore(person),
          };
        } catch (mappingError) {
          console.error('Error mapping prospect:', mappingError, person);
          return null;
        }
      }).filter(Boolean); // Remove any null entries from mapping errors

      const totalResults = response.data.pagination?.total_entries || prospects.length;

      console.log(`📊 Apollo search completed: ${prospects.length} prospects mapped, ${totalResults} total available`);

      return {
        results: prospects,
        total: totalResults,
      };
    } catch (error: any) {
      console.error('❌ Apollo API error:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status,
        config: {
          url: error.config?.url,
          method: error.config?.method,
          headers: error.config?.headers,
        }
      });

      // If it's a 422 error, log the specific validation issues
      if (error.response?.status === 422) {
        console.error('💥 Apollo API validation error:', JSON.stringify(error.response.data, null, 2));
      }

      // If it's a 401 error, the API key might be invalid
      if (error.response?.status === 401) {
        console.error('🔑 Apollo API authentication failed - check API key');
      }

      // If it's a 403 error, might be rate limited or quota exceeded
      if (error.response?.status === 403) {
        console.error('🚫 Apollo API access denied - check quota and rate limits');
      }

      return { results: [], total: 0 };
    }
  }

  private formatCompanySize(size: any): string {
    if (typeof size === 'string') return size;
    if (typeof size === 'number') return size.toString();
    return '';
  }

  private formatLocation(organization: any): string {
    if (!organization) return '';
    
    const parts = [];
    if (organization.city) parts.push(organization.city);
    if (organization.state) parts.push(organization.state);
    if (organization.country) parts.push(organization.country);
    
    return parts.join(', ') || organization.location || '';
  }

  private calculateConfidenceScore(person: any): number {
    let score = 0.5; // Base score
    
    // Boost score based on data quality
    if (person.email) score += 0.2;
    if (person.email_verified_at) score += 0.1;
    if (person.phone_numbers && person.phone_numbers.length > 0) score += 0.1;
    if (person.linkedin_url) score += 0.1;
    if (person.organization?.name) score += 0.1;
    
    return Math.min(score, 1.0);
  }

  private mapCompanySize(size?: string): string[] {
    // Direct mapping for frontend values
    const directMap: Record<string, string[]> = {
      '1-10': ['1-10'],
      '11-50': ['11-50'],
      '51-200': ['51-200'],
      '200+': ['201-500', '501-1000', '1001-5000', '5001-10000', '10001+'],
      'all': [],
    };

    // Legacy mapping for backward compatibility
    const sizeMap: Record<string, string[]> = {
      'startup': ['1-10', '11-50'],
      'smb': ['51-200', '201-500'],
      'midmarket': ['501-1000', '1001-5000'],
      'enterprise': ['5001-10000', '10001+'],
    };

    return directMap[size || ''] || sizeMap[size || ''] || [];
  }
}

File: backend/services/apollo_client.py
"""
Apollo.io API client for prospect discovery
"""
import os
import httpx
import logging
from typing import Dict, Any, Optional, List

logger = logging.getLogger(__name__)

class ApolloClient:
    """Apollo.io API client for prospect search and discovery"""
    
    def __init__(self):
        self.api_key = os.getenv('APOLLO_API_KEY')
        self.base_url = 'https://api.apollo.io/v1'
        self.client = httpx.AsyncClient(timeout=30.0)
        
        if not self.api_key:
            logger.warning("Apollo API key not configured - some features may be limited")
    
    async def search_people(self, params: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Search for people using Apollo.io API"""
        if not self.api_key:
            logger.warning("Apollo API key not available")
            return None
        
        try:
            headers = {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
                'X-Api-Key': self.api_key
            }
            
            # Build request payload
            payload = {
                'q_keywords': params.get('q_keywords', ''),
                'page': params.get('page', 1),
                'per_page': min(params.get('per_page', 50), 200)
            }
            
            # Add organization filters
            if params.get('organization_industry_tag_names'):
                payload['organization_industry_tag_names'] = params['organization_industry_tag_names']
            
            if params.get('organization_num_employees_ranges'):
                payload['organization_num_employees_ranges'] = params['organization_num_employees_ranges']
            
            # Add person filters
            if params.get('person_locations'):
                payload['person_locations'] = params['person_locations']
            
            if params.get('person_titles'):
                payload['person_titles'] = params['person_titles']
            
            logger.info(f"🔍 Apollo API request: {self.base_url}/mixed_people/search")
            
            response = await self.client.post(
                f"{self.base_url}/mixed_people/search",
                json=payload,
                headers=headers
            )
            
            if response.status_code == 200:
                result = response.json()
                logger.info(f"✅ Apollo search successful: {len(result.get('people', []))} results")
                return result
            else:
                logger.error(f"❌ Apollo API error {response.status_code}: {response.text}")
                return None
                
        except Exception as e:
            logger.error(f"❌ Apollo API request failed: {str(e)}")
            return None
    
    async def get_person(self, person_id: str) -> Optional[Dict[str, Any]]:
        """Get detailed person information by ID"""
        if not self.api_key:
            return None
        
        try:
            headers = {
                'Content-Type': 'application/json',
                'X-Api-Key': self.api_key
            }
            
            response = await self.client.get(
                f"{self.base_url}/people/{person_id}",
                headers=headers
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Apollo person fetch error {response.status_code}: {response.text}")
                return None
                
        except Exception as e:
            logger.error(f"Apollo person fetch failed: {str(e)}")
            return None
    
    async def enrich_person(self, email: str) -> Optional[Dict[str, Any]]:
        """Enrich person data by email"""
        if not self.api_key:
            return None
        
        try:
            headers = {
                'Content-Type': 'application/json',
                'X-Api-Key': self.api_key
            }
            
            payload = {
                'email': email,
                'reveal_personal_emails': True
            }
            
            response = await self.client.post(
                f"{self.base_url}/people/match",
                json=payload,
                headers=headers
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Apollo enrichment error {response.status_code}: {response.text}")
                return None
                
        except Exception as e:
            logger.error(f"Apollo enrichment failed: {str(e)}")
            return None
    
    def is_available(self) -> bool:
        """Check if Apollo API is available"""
        return bool(self.api_key)
    
    async def close(self):
        """Close the HTTP client"""
        await self.client.aclose()